lib/resources/grub_conf.rb

```diff
@@ -71,35 +73,74 @@ def to_s
   ######################################################################
 
   def grub2_parse_kernel_lines(content, conf)
-    # Find all "menuentry" lines and then parse them into arrays
-    menu_entry = 0
+    menu_entries = extract_menu_entries(content)
+
+    if @kernel == 'default'
+      default_menu_entry(menu_entries, conf['GRUB_DEFAULT'])
+    else
+      menu_entries.select { |entry| entry['name'] == @kernel }[0]
+    end
+  end
+
+  def extract_menu_entries(content)
+    menu_entries = []
+
     lines = content.split("\n")
-    kernel_opts = {}
-    kernel_opts['insmod'] = []
-    lines.each_with_index do |file_line, index|
-      next unless file_line =~ /(^|\s)menuentry\s.*/
-      lines.drop(index+1).each do |kernel_line|
-        next if kernel_line =~ /(^|\s)(menu|}).*/
-        if menu_entry == conf['GRUB_DEFAULT'].to_i && @kernel == 'default'
-          if kernel_line =~ /(^|\s)initrd.*/
-            kernel_opts['initrd'] = kernel_line.split(' ')[1]
-          end
-          if kernel_line =~ /(^|\s)linux.*/
-            kernel_opts['kernel'] = kernel_line.split
-          end
-          if kernel_line =~ /(^|\s)set root=.*/
-            kernel_opts['root'] = kernel_line.split('=')[1].tr('\'', '')
-          end
-          if kernel_line =~ /(^|\s)insmod.*/
-            kernel_opts['insmod'].push(kernel_line.split(' ')[1])
-          end
-        else
-          menu_entry += 1
-          break
+    lines.each_with_index do |line, index|
+      next unless line =~ /^menuentry\s+.*/
+      entry = {}
+      entry['insmod'] = []
+
+      # Extract name from menuentry line
+      capture_data = line.match(/(?:^|\s+).*menuentry\s*['|"](.*)['|"]\s*--/)
+
+      if capture_data.captures[0]
```

Comment:
I still think it's possible for `capture_data` to POSSIBLY return nil, and then this line will throw a "undefined method on NilClass" error.

If you really want to fail the whole resource if we can't parse the line, this should probably look like this:

```ruby
if capture_data.nil? || capture_data.captures[0].nil?
  raise Inspec::Exceptions::ResourceFailed "Cannot extract menuentry name from line: #{line}"
end

entry['name'] = capture_data.captures[0]
```

You could also use the safe navigation operator to your advantage here as well:

```ruby
entry['name'] = capture_data&.captures[0]
if entry['name'].nil?
  raise Inspec::Exceptions::ResourceFailed "Cannot extract menuentry name from line: #{line}"
end
```

Reply:
I agree, better to raise anything but "undefined method for NilClass".

I went with your former suggestion. The safe navigation operator is neat, but it threw me for a loop the first time I saw it and I wouldn't want it to trip anyone else up later.